File: root-cpp-java-performance/readOneBranch.cxx
#include <iostream>
#include <time.h>

#include "TFile.h"
#include "TTree.h"
#include "TLeaf.h"
#include "TBranch.h"
#include "TClass.h"
#include "TBranchElement.h"
#include "TCollection.h"
#include "TSystem.h"

void runfile(const char* fileName) {
  std::cout << fileName << std::endl;

  double total = 0.0;

  TFile *tfile = new TFile(fileName);
  TTree *ttree;
  tfile->GetObject("Events", ttree);

  std::cout << "number of TBaskets ???" << std::endl;

  // get the branch with a full path
  // it's a TBranchElement because recoTracks_generalTracks__RECO.obj is a structure
  // flat ntuples don't have TBranchElements, but we only need TBranchElement::GetBranchCount for structures
  TBranchElement *branch = (TBranchElement*)ttree->GetBranch("recoTracks_generalTracks__RECO.obj.chi2_");

  // essential!!! MakeClass mode lets us view the structure one leaf at a time.
  ttree->SetMakeClass(1);

  // has to be reassigned for each new TFile (we only know the maximum for *this file*)
  int bufferSize = ((TLeaf*)(branch->GetListOfLeaves()->First()))->GetLeafCount()->GetMaximum();

  // allocate a buffer that's just big enough
  Float_t *buffer = new Float_t[bufferSize];

  // have the branch fill this buffer
  ttree->SetBranchAddress("recoTracks_generalTracks__RECO.obj.chi2_", buffer);

  // allocating a place to put the size is also essential
  Int_t size = 0;
  ttree->SetBranchAddress("recoTracks_generalTracks__RECO.obj",&size);

  // clock stuff
  clock_t lastTime = clock();    
  long items = 0L;
  long itemsPerPrint = 100000L;

  // the loop over events
  Long64_t numEvents = ttree->GetEntries();
  for (Long64_t i = 0;  i < numEvents;  i++) {
    // essential! GetEntry from the branch, not the ttree
    branch->GetEntry(i);

    // get the number of elements (tracks in this case)
    int numTracks = branch->GetBranchCount()->GetNdata();

    // and loop over them
    for (int j = 0;  j < numTracks;  ++j) {
      // getting the data does not involve any function calls
      total += buffer[j];

      // clock stuff
      items++;
      if (items % itemsPerPrint == 0) {
        clock_t now = clock();
        std::cout << 1.0 * (now - lastTime) / itemsPerPrint / CLOCKS_PER_SEC * 1e9 << " ns/item" << std::endl;
        lastTime = now;
      }
    }
  }

  // clock and checksum results
  std::cout << "check total " << total << " == 1.55104e+07 (" << (abs(total - 1.55104e+07) < 1e-6*1.55104e+07 ? "true" : "false") << ")" << std::endl;
}

void readOneBranch() {
  runfile("Mu_Run2010B-Apr21ReReco-v1_AOD.root");
  runfile("copy2.root");
  runfile("copy3.root");
  runfile("copy4.root");
  runfile("copy5.root");
  gSystem->Exit(0);
}


// SetMakeClass(1)

// GetLenType() * GetLenStatic() * GetLeafCount()->GetMaximum()

File: root-cpp-java-performance/ReadOneBranch.java
import hep.io.root.core.RootInput;
import hep.io.root.*;
import hep.io.root.interfaces.*;

public class ReadOneBranch {
    final static long itemsPerPrint = 100000L;
    static long lastTime = 0L;
    static long items = 0;

    private static void printout() {
        long now = System.nanoTime();
        System.out.println(String.format("%g ns/item", 1.0 * (now - lastTime) / itemsPerPrint));
        lastTime = now;
    }

    public static void main(String[] args) throws java.io.IOException, RootClassNotFound {
        String[] fileNames = {"Mu_Run2010B-Apr21ReReco-v1_AOD.root", "copy2.root", "copy3.root", "copy4.root", "copy5.root"};
        for (String fileName : fileNames) {
            System.out.println(fileName);

            double total = 0.0;

            RootFileReader reader = new RootFileReader(fileName);
            TTree tree = (TTree)reader.get("Events");

            TBranch branch = tree.getBranch("recoTracks_generalTracks__RECO.").getBranchForName("obj").getBranchForName("chi2_");
            TLeaf leaf = (TLeaf)branch.getLeaves().get(0);

            long[] startingEntries = branch.getBasketEntry();
            System.out.println(String.format("number of TBaskets %d", startingEntries.length - 1));
            
            lastTime = System.nanoTime();
            items = 0;

            for (int i = 0;  i < startingEntries.length - 1;  i++) {
                long endEntry = startingEntries[i + 1];

                // all but the last one
                for (long entry = startingEntries[i];  entry < endEntry - 1;  entry++) {
                    RootInput in = branch.setPosition(leaf, entry + 1);
                    long endPosition = in.getPosition();
                    in = branch.setPosition(leaf, entry);
                    while (in.getPosition() < endPosition) {
                        total += in.readFloat();

                        items += 1;
                        if (items % itemsPerPrint == 0) printout();
                    }
                }

                // the last one
                RootInput in = branch.setPosition(leaf, endEntry - 1);
                long endPosition = in.getLast();
                while (in.getPosition() < endPosition) {
                    total += in.readFloat();

                    items += 1;
                    if (items % itemsPerPrint == 0) printout();
                }
            }

            System.out.println(String.format("check total %g == 1.55104e+07 (%s)", total, Math.abs(total - 1.55104e+07) > 1e-12*1.55104e+07 ? "true" : "false"));
        }
    }
}
